### ISSUE TITLE
Parallel simulations with differing random seed

### ISSUE THREAD

**Issue by davestanley** [DynaSim/DynaSim#297]
I discovered that, even with random seed set to shuffle, DynaSim does not always shuffle the random seed when in parallel mode.

At least, when I ran 8 simulations, it seems it used a single random seed for the first 2 and then a different random seed for the remainder. I've attached a screenshot (note the first 2 traces are the same, as are the last 6).

I'm going to look in to fixing this.

![image](https://user-images.githubusercontent.com/5439305/33641008-acbd2506-da01-11e7-9669-6146bcbac9ed.png)

**Comment by davestanley** [comment]
From looking at the solve files, there are 2 blocks of code where constants are set: (1) Fixed variables and (2) Initial conditions. The random seed is set in the second one (despite there being calls to unifrnd and others in the first block).

@jsherfey is there any reason we can't move this call to `rng` up from the Initial conditions section to the Fixed Variables section?

```
% ------------------------------------------------------------
% Fixed variables:
% ------------------------------------------------------------
NG_iPeriodicPulsesiSYNNested_Tend = T(end);
NG_iPeriodicPulsesiSYNNested_PPgsyn0 = unifrnd(p.NG_iPeriodicPulsesiSYNNested_PP_gSYN*(1-p.NG_iPeriodicPulsesiSYNNested_PP_g_SYN_hetero),p.NG_iPeriodicPulsesiSYNNested_PP_gSYN*(1+p.NG_iPeriodicPulsesiSYNNested_PP_g_SYN_hetero),[p.NG_Npop 1])';
NG_iPeriodicPulsesiSYNNested_PPgsyn = NG_iPeriodicPulsesiSYNNested_PPgsyn0.* double(NG_iPeriodicPulsesiSYNNested_PPgsyn0 > 0);
NG_iPeriodicPulsesiSYNNested_s0 =  getDeltaTrainPresets2(p.NG_iPeriodicPulsesiSYNNested_PPfreq,p.NG_iPeriodicPulsesiSYNNested_PPshift,NG_iPeriodicPulsesiSYNNested_Tend,p.NG_iPeriodicPulsesiSYNNested_dt,p.NG_iPeriodicPulsesiSYNNested_ap_pulse_num,p.NG_iPeriodicPulsesiSYNNested_ap_pulse_delay,p.NG_iPeriodicPulsesiSYNNested_pulse_train_preset);
NG_iPeriodicPulsesiSYNNested_s0_mask =  getDeltaTrainPresets2(p.NG_iPeriodicPulsesiSYNNested_PPmaskfreq,p.NG_iPeriodicPulsesiSYNNested_PPmaskshift,NG_iPeriodicPulsesiSYNNested_Tend,p.NG_iPeriodicPulsesiSYNNested_dt,0,0,0);
```

... (more code) ...

```
% ------------------------------------------------------------
% Initial conditions:
% ------------------------------------------------------------
% seed the random number generator
rng(p.random_seed);
t=0; k=1;

% STATE_VARIABLES:
NG_V_last = -65*ones(1,p.NG_Npop);
NG_V = zeros(nsamp,p.NG_Npop);
  NG_V(1,:) = NG_V_last;
NG_iPeriodicPulsesiSYNNested_s_last =  p.NG_iPeriodicPulsesiSYNNested_IC*ones(1,p.NG_Npop)+p.NG_iPeriodicPulsesiSYNNested_IC_noise.*rand(1,p.NG_Npop);
NG_iPeriodicPulsesiSYNNested_s = zeros(nsamp,p.NG_Npop);
```

**Comment by jsherfey** [comment]
No, it should be moved up.



On Dec 5, 2017 9:31 PM, "Dave" <notifications@github.com> wrote:

From looking at the solve files, there are 2 blocks of code where constants
are set: (1) Fixed variables and (2) Initial conditions. The random seed is
set in the second one (despite there being calls to unifrnd and others in
the first block).

@jsherfey <https://github.com/jsherfey> is there any reason we can't move
this call to rng up from the Initial conditions section to the Fixed
Variables section?

% ------------------------------------------------------------
% Fixed variables:
% ------------------------------------------------------------
NG_iPeriodicPulsesiSYNNested_Tend = T(end);
NG_iPeriodicPulsesiSYNNested_PPgsyn0 =
unifrnd(p.NG_iPeriodicPulsesiSYNNested_PP_gSYN*(1-p.NG_iPeriodicPulsesiSYNNested_PP_g_SYN_hetero),p.NG_iPeriodicPulsesiSYNNested_PP_gSYN*(1+p.NG_iPeriodicPulsesiSYNNested_PP_g_SYN_hetero),[p.NG_Npop
1])';
NG_iPeriodicPulsesiSYNNested_PPgsyn =
NG_iPeriodicPulsesiSYNNested_PPgsyn0.*
double(NG_iPeriodicPulsesiSYNNested_PPgsyn0 > 0);
NG_iPeriodicPulsesiSYNNested_s0 =
getDeltaTrainPresets2(p.NG_iPeriodicPulsesiSYNNested_PPfreq,p.NG_iPeriodicPulsesiSYNNested_PPshift,NG_iPeriodicPulsesiSYNNested_Tend,p.NG_iPeriodicPulsesiSYNNested_dt,p.NG_iPeriodicPulsesiSYNNested_ap_pulse_num,p.NG_iPeriodicPulsesiSYNNested_ap_pulse_delay,p.NG_iPeriodicPulsesiSYNNested_pulse_train_preset);
NG_iPeriodicPulsesiSYNNested_s0_mask =
getDeltaTrainPresets2(p.NG_iPeriodicPulsesiSYNNested_PPmaskfreq,p.NG_iPeriodicPulsesiSYNNested_PPmaskshift,NG_iPeriodicPulsesiSYNNested_Tend,p.NG_iPeriodicPulsesiSYNNested_dt,0,0,0);

... (more code) ...

% ------------------------------------------------------------
% Initial conditions:
% ------------------------------------------------------------
% seed the random number generator
rng(p.random_seed);
t=0; k=1;

% STATE_VARIABLES:
NG_V_last = -65*ones(1,p.NG_Npop);
NG_V = zeros(nsamp,p.NG_Npop);
  NG_V(1,:) = NG_V_last;
NG_iPeriodicPulsesiSYNNested_s_last =
p.NG_iPeriodicPulsesiSYNNested_IC*ones(1,p.NG_Npop)+p.NG_iPeriodicPulsesiSYNNested_IC_noise.*rand(1,p.NG_Npop);
NG_iPeriodicPulsesiSYNNested_s = zeros(nsamp,p.NG_Npop);

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/DynaSim/DynaSim/issues/297#issuecomment-349511248>, or mute
the thread

**Comment by davestanley** [comment]
Just submit a fix to this in #297. I'm not so familiar with these functions, so please have a look. Basically, I just duplicated the random seed code in the Fixed variables section.  @jsherfey

**Comment by DynaSim** [comment]
That looks ok to me.
